### PR TITLE
WIP: remove distributed loading, disentangle loading from source search

### PR DESCRIPTION
**Pull request by StefanKarpinski**
Change `include` to only ever load from the current node. When loading files via `using` and `import`, don't look for a literal file path. That kind of loading only belongs in `include`, which loads files, not modules. These two ways of loading code used to be conflated but these days, they're very clearly separated – this disentangles those unrelated code paths.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Does this supersede #19866?

File: base/deprecated.jl [hunk 1]
```diff
@@ -1499,4 +1499,13 @@ end
...
 @deprecate promote_eltype_op(op, As...) promote_op(op, map(eltype, As)...)
 
+function include(_path::AbstractString...)
+    depwarn("multi-argument `include` is deprecated, use multiple `include` calls instead", :include)
+    local result
+    for path in _path
+        result = include(path)
+    end
+    result
+end
```
Comment (ararslan) [inline base/deprecated.jl]: Perhaps the deprecation of multi-argument `include` should be in #19909 instead?

**Comment by stevengj** [comment]
The reference to "node 1" in the `include_string` docstring should be removed.  (The `include_string` docstring should also explain its 2nd argument.)

**Comment by stevengj** [comment]
So after this PR, how do you load code on a worker node?

**Comment by StefanKarpinski** [comment]
This is part of https://github.com/JuliaLang/julia/pull/19866 – I'm building a cascade of changes so that we can merge parts as they're ready. The idea is to treat code loading just like anything else: write `@everywhere using Foo` to load `Foo` everywhere and loading is done from each machine locally. https://github.com/JuliaLang/julia/pull/19866 will add enough extensibility that loading code from node 1 can be an opt-in extension for people who need it.

File: base/loading.jl [hunk 2]
```diff
@@ -90,44 +90,15 @@ function try_path(prefix::String, base::String, name::String)
...
-    end
-    if wd !== nothing
-        isfile_casesensitive(joinpath(wd,name)) && return joinpath(wd,name)
-    end
-    p = try_path(Pkg.dir(), base, name)
-    p !== nothing && return p
-    for prefix in LOAD_PATH
+function find_in_path(base::String)
+    name = "$base.jl"
+    for prefix in [Pkg.dir(); LOAD_PATH]
```
Comment (tkelman) [inline base/loading.jl]: ref https://github.com/JuliaLang/julia/pull/15934#r60162030, this concatenation may still be expensive?

Comment (StefanKarpinski) [inline base/loading.jl]: Yeah, I'm sure it is. Probably should just leave it the way it was – I mostly wanted to make it clearer here that the code behaves as if `Pkg.dir()` was in `LOAD_PATH`.

**Comment by stevengj** [comment]
I'm just worried about distributed systems where only node 1 has access to the filesystem where Julia is installed.  Or is this something that ClusterManagers.jl could add as an extension that would work transparently?

**Comment by StefanKarpinski** [comment]
The vague plan is to add support for "virtual filesystems" along the lines of something like `RemotePath(node, "/path/to/remote/directory")`, make sure that all of the code loading infrastructure can load code from this kind of virtualized remote filesystem object, and then allow putting these into `LOAD_PATH` (or whatever replaces it). Thus, you would opt into loading from node 1 by doing `unshift!(LOAD_PATH, RemotePath(1, path))`.

**Comment by stevengj** [comment]
It's a bit worrisome to remove this functionality from 0.6, since the impending feature freeze makes it seem likely that a replacement functionality will not be available until the next release.

**Comment by StefanKarpinski** [comment]
Yes, we should have gotten this in earlier. Unfortunately, we need some kind of hooking mechanism in here if Pkg3 is going to be usable with 0.6 as an external package.

**Comment by amitmurthy** [comment]
Leaving a comment in order to be notified on this conversation.

Is there a less noisy way to do this?

**Comment by ararslan** [comment]
@amitmurthy You can click "Subscribe" on the right side panel beneath the milestone, labels, etc.

**Comment by amitmurthy** [comment]
@ararslan , I was looking for email notifications. The "Subscribe" button enables notifications on Github web interface. In user settings there is a differentiation between "participating" and "watching".

**Comment by amitmurthy** [comment]
@stefan, asking this as you have removed the 0.6 milestone tag from this issue.

We can still try and get #19073 in. Are you looking to change "loading.jl" beyond #20008 now?

**Pull request closed without merging**